### ISSUE TITLE
Adding support for Focal Loss?

### ISSUE THREAD

**Issue by paganpasta** [google-deepmind/optax#525]
Hi,

Thanks for the amazing repo. I have been using it for performing experiments focusing mostly on computer vision tasks.

I currently use [focal_loss](https://paperswithcode.com/method/focal-loss) in my work. I was wondering if adding the loss to this repo would be something useful.

I currently use it in the following manner

```python3
def focal_loss(
    logits: chex.Array,
    labels: chex.Array,
    alpha: chex.Array,
    gamma: float = 2.0,
) -> chex.Array:
  """Computes the focal loss between sets of logits and labels.

  Computes the focal loss

  References:
    [Lin et al, 2017](https://arxiv.org/abs/1708.02002)

  Args:
    logits: Unnormalized log probabilities, with shape `[..., num_classes]`.
    labels: Valid probability distributions (non-negative, sum to 1), e.g a
      one hot encoding specifying the correct class for each input;
      must have a shape broadcastable to `[..., num_classes]`.
    alpha: Weighting factor, with shape `[num_classes]` and values in the range `[0, 1]`.
    gamma: Focusing parameter `>=0`. It controls the contribution of higher confidence predictions.
      Defaults to 2.

  Returns:
    focal loss between each prediction and the corresponding target
    distributions, with shape `[...]`.
  """
  chex.assert_type([logits], float)
  focus = jnp.power(-jax.nn.softmax(logits, axis=-1) + 1.0, gamma)
  loss = -labels * alpha * focus * jax.nn.log_softmax(logits, axis=-1)
  return jnp.sum(loss, axis=-1)
```

**Comment by mtthss** [comment]
Apologies for the delay, this is a great suggestion, can you put together a PR?

**Comment by v0lta** [comment]
I played with this a little the alpha-parameter has a problem in the current version. If it indeed has the shape [num_classes] the function crashes.

The current implementation also differs from the implementation in:
https://github.com/facebookresearch/fvcore/blob/0f2b23b6f93e36041d9a74764ee824541cf0a0e5/fvcore/nn/focal_loss.py#L41

I suggest to change the function as follows:
```
def focal_loss(
    logits: jnp.ndarray,
    labels: jnp.ndarray,
    alpha: float = 0.25,
    gamma: float = 2.0,
) -> jnp.ndarray:
    """Compute the focal loss between sets of logits and labels.

    Computes the focal loss

    References:
      [Lin et al, 2017](https://arxiv.org/abs/1708.02002)

    Args:
      logits: Unnormalized log probabilities, with shape `[..., num_classes]`.
      labels: Valid probability distributions (non-negative, sum to 1), e.g a
        one hot encoding specifying the correct class for each input;
        must have a shape broadcastable to `[..., num_classes]`.
      alpha (float): Weighting factor in range (0,1) to balance
          positive vs negative examples or -1 for ignore. Default: ``0.25``
      gamma: Focusing parameter `>=0`. It controls the contribution of higher confidence predictions.
        Defaults to 2.

    Returns:
      focal loss between each prediction and the corresponding target
      distributions, with shape `[...]`.
    """
    focus = jnp.power(-jax.nn.softmax(logits, axis=-1) + 1.0, gamma)
    loss = -labels * focus * jax.nn.log_softmax(logits, axis=-1)

    if alpha >= 0:
        alpha_t = alpha * labels + (1 - alpha) * (1 - labels)
        loss = alpha_t * loss
    return jnp.sum(loss, axis=-1)
```

Which mirrors the original authors' implementation.

**Comment by v0lta** [comment]
The original code does not use a softmax, I think this might be confusing. Perhaps this can be reflected in the name somehow?

**Comment by v0lta** [comment]
See https://github.com/deepmind/optax/pull/574 for a sigmoid version.

**Comment by v0lta** [comment]
I am not sure the computation of the focus term is currently correct. See i.e. https://github.com/artemmavrin/focal-loss/blob/7a1810a968051b6acfedf2052123eb76ba3128c4/src/focal_loss/_binary_focal_loss.py#L512  .
I think the key idea is to expand equation two from https://arxiv.org/pdf/1708.02002v2.pdf . I think the repository link above does something similar since the softmax gives us probabilities that sum to one.

**Comment by paganpasta** [comment]
Hi @v0lta 

Can you share the reason for the crash? There is a unit test  that I have added in the PR, are you using as described there? 
We can use scalar/vector for alpha, but, I have to emphasize that this is a multi-class focal loss and not a binary one. That is why it utilises softmax instead of sigmoid. 
As mentioned in the documentation string, it takes as input a `[..., num_classes]` object. It could be that you tried for num_classes=1 and it didn't work.

**Comment by v0lta** [comment]
- You are right. Something like equation two from Lin et al., 2017, does not exist in the multi-class version. Bishops equation (5.24) confirms this [1].
- Regarding the error, I think I had a shape that was unequal to [num_classes] it first. So that is a problem on my end. I will remove the comment from the pull request.

[1]  Pattern Recognition and Machine Learning, Christopher M. Bishop .

**Comment by fabianp** [comment]
Current pull request: https://github.com/google-deepmind/optax/pull/599

**Comment by fabianp** [comment]
Closing this issue since we merged #599 . Please don't hesitate to open a new issue if you have further comments on this implementation
